speed up date time parsing

shashi opened the pull request:
This PR builds on @simonbyrne's #18000 and makes the (120x :laughing:) speed up available for custom date formats. It also re-does the DateFormat type to be sufficiently parameterized so that an efficient parsing function can be "generated". This also incidentally makes it easier to add new features to the parser.

before:
```julia
julia> @benchmark DateTime("2010-10-10T10:10:10.10")
BenchmarkTools.Trial: 
  memory estimate:  3.11 kb
  allocs estimate:  85
  --------------
  minimum time:     20.125 μs (0.00% GC)
  median time:      21.233 μs (0.00% GC)
  mean time:        21.818 μs (1.38% GC)
  maximum time:     3.128 ms (96.19% GC)
  --------------
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

after:
```julia
julia> @benchmark DateTime("2010-10-10T10:10:10.10")
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     164.320 ns (0.00% GC)
  median time:      168.183 ns (0.00% GC)
  mean time:        168.666 ns (0.00% GC)
  maximum time:     291.389 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     750
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

Also works fast for custom date format strings.

```julia
julia> @benchmark DateTime("01-01-2010X53:23:12", dateformat"dd-mm-YXS:M:H")
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     148.979 ns (0.00% GC)
  median time:      151.564 ns (0.00% GC)
  mean time:        153.853 ns (0.00% GC)
  maximum time:     327.571 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     818
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

fixes #15888, #13644
supersedes #18000

simonbyrne committed:
Custom parser for dates.

tkelman commented on base/exports.jl:
```diff
@@ -1009,6 +1009,8 @@ export
     Date,
     DateTime,
+    DateFormat,
+    @dateformat_str,
```
should be added to the manual index if exported

tkelman replied:
see contributing.md

shashi replied:
If you meant I need to add this to the manual, I have. I was wondering if there's some kind of list I need to update when you said "manual index"

tkelman replied:
not talking about doc/manual, talking about doc/stdlib for the docstring. yes there is a list, again see contributing.md

shashi replied:
Ah, okay. Will do.

StefanKarpinski commented:
Amazing work, @shashi! I would note that this *still* doesn't preclude having even more efficient custom code for the three or so common date formats.

shashi commented:
@StefanKarpinski that's right, and you don't even have to give them a name and add them to a type hierarchy. :) Just dispatch on `typeof(dateformat"mycommonformat")`.

ViralBShah commented:
Can this be backported to 0.5?

shashi commented:
It breaks compatibility with the previous `DateFormat` type (presumably DateFormat was an implementation detail) and changes parsing behavior in some corner cases (see diff of [tests](https://github.com/JuliaLang/julia/pull/19545/files#diff-3518688434c17041fea5e10ad667d7e9))...

tkelman commented on test/dates/io.jl:
```diff
@@ -268,7 +275,7 @@ f = "y m d"
 @test_throws ArgumentError Dates.Date(" 1 1 32",f)
 @test_throws ArgumentError Dates.Date("# 1 1 32",f)
 # can't find 1st space delimiter,s o fails
```
comment no longer applies?

omus commented on base/dates/types.jl:
```diff
@@ -137,6 +137,18 @@ daysinmonth(y,m) = DAYSINMONTH[m] + (m == 2 && isleapyear(y))
+    rata = xs[7] + 1000*(xs[6] + 60*xs[5] + 3600*xs[4] + 86400*Base.Dates.totaldays(xs[1],xs[2],xs[3]))
+    return DateTime(Base.Dates.UTM(rata))
+end
+
```
What's the advantage of having this duplicate constructor versus:
```julia
DateTime(t::NTuple{7,Int}) = DateTime(t[1], t[2], t[3], t[4], t[5], t[6], t[7])
```

omus replied:
Current implementation will run into issues with integer overflow on 32-bit systems

simonbyrne replied:
Or just `DateTime(t...)`

omus commented on base/dates/io-util.jl:
```diff
@@ -0,0 +1,178 @@
+        return R(reorder_args(parts, fmt.field_order, fmt.field_defaults, err_idx)::NTuple{7,Int})
+
+        @label error
+        return R((err_idx,state,0,0,0,0,0), true)
```
If you want to indicate `null` here you want `false`

simonbyrne commented on base/dates/io-util.jl:
```diff
@@ -0,0 +1,178 @@
+    end
+end
+
+_create_timeobj(tup, T::Type{DateTime}) = T(tup)
```
why not `T(tup...)`?

shashi replied:
I fixed changed this in a later commit. this review business is a bit confusing. Thanks for reviewing!

simonbyrne commented:
This looks pretty awesome. A few minor changes mostly pointed out by @omus.

omus commented:
I am concerned about these changes as they break support for custom parser slots which don't return integers. Specifically, these new changes won't work with parsing TimeZone types.

vtjnash commented on base/dates/io.jl:
```diff
@@ -1,29 +1,97 @@
+is used. This object is passed as the last argument to
+`tryparsenext` and `format` defined for each `AbstractDateToken` type.
+"""
+immutable DateLocale{lang} end
```
I strongly recommend against baking `lang` into a type parameter. This will actually make the API more awkward to extend than presenting it to the user as a dict.

shashi replied:
Okay. I can make this type hold two dictionaries.

simonbyrne replied:
We may want to think about how this can be generalised to other locale-based print functions (e.g. decimal separators), though that may be better in a different PR.

vtjnash commented on base/dates/io-util.jl:
```diff
@@ -0,0 +1,178 @@
+        @label error
+        return R((err_idx,state,0,0,0,0,0), true)
+    end
+end
```
I know the classic way of doing it in a static language (hardcoding a jump table / switch / vtable), but afaik there isn't really a good way to express that optimization in Julia right now (other than by doing a generated function like this). fwiw, since this only uses `@nexpr`, it appears to be a completely valid generated function.

omus committed:
Refactor DateFormat, Delim and DatePart types

Combination of commits by Chris Vogt:

- Minor refactor to Delim
- Refactor DatePart
- Fixes inefficiency by removing need for tryparsenext_base10_frac
- Remove need for tryparsenext_base10_frac
- Performance is now consistent for millisecond width changes. Otherwise
- performance seems to be about the same.
- Fix DateTime tuple constructor
- On 32-bit systems the UTM calculation could run into integer overflow.
- Bugfixes

omus committed:
Making interface easier to work with

omus committed:
Performance tweaks

1. Reduce bounds checks
2. Inline max_width, min_width, and _create_object
3. Remove dead tryparse method

Getting performance back up

Remove unnecessary method

Improving performance futher

Avoiding indexing into tuple

simonbyrne commented:
@nanosoldier `runbenchmarks("dates", vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19545#issuecomment-267093623) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/1332c7ac32db90881a48b73dddb4496e88919b5c/248f937_vs_bfc72b6/report.md). cc @jrevels

davidanthoff commented:
Might it make sense to do #16928 as part of this as well?

omus committed:
Refactor parse functions

Makes code clearer and more performant.

1. Refactor to be more like Int parsing code
2. remove unreadable @chk1 macro
3. parse can now take output type as first argument removing the need
for it in DateFormat

Refactor to be more like Int parsing code

Remove `@chk1` as it was only used once

Removed need for result type in DateFormat

omus commented:
I'm refactoring @shashi's code to clean it up and maybe find some additional performance tweaks (I've found a couple). I'll be update #19519 with this work unless someone feels I should open a new PR and close #19519.

ViralBShah commented:
I don't see any reason to open a new PR.

simonbyrne commented:
Interesting: I wonder why the printing is now slower?

shashi commented:
@simonbyrne I did not benchmark printing speeds, but had to redo it to accomodate the new DateFormat. There should be easy speedups available.

JeffBezanson commented:
This is some excellent work. Real julia wizardry on display here!

shashi commented:
Thanks @JeffBezanson :smile:

shashi commented:
@omus I have merged your changes in here with the required changes.

omus committed:
Changes to base/dates/query.jl for clairity

omus committed:
Rename base/dates/io-util.jl to base/dates/parse.jl

omus committed:
Return string from tryparsenext_word. Validation for day of week.

tryparsenext_word now returns the word

Modification does slightly impact performance when using 'u' or 'U' as
it creates an unused SubString but the interface is much more consistent
and it allows us to verify that the day of week is valid.

Performance test:

before

```julia
julia> @benchmark DateTime("Sat, 12 Nov 2016 07:45:36", Base.Dates.RFC1123Format)
BenchmarkTools.Trial:
  memory estimate:  384.00 bytes
  allocs estimate:  6
  --------------
  minimum time:     455.030 ns (0.00% GC)
  median time:      478.150 ns (0.00% GC)
  mean time:        571.465 ns (7.06% GC)
  maximum time:     19.411 μs (93.96% GC)
  --------------
  samples:          10000
  evals/sample:     197
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

after

```julia
julia> @benchmark DateTime("Sat, 12 Nov 2016 07:45:36", Base.Dates.RFC1123Format)
BenchmarkTools.Trial:
  memory estimate:  528.00 bytes
  allocs estimate:  12
  --------------
  minimum time:     536.836 ns (0.00% GC)
  median time:      562.124 ns (0.00% GC)
  mean time:        682.729 ns (8.93% GC)
  maximum time:     20.317 μs (95.27% GC)
  --------------
  samples:          10000
  evals/sample:     189
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

Parsing invalid days of week now raise exception

Has a negative impact on parsing dates using day of week.

Improve performance on word slots

We can get a serious performance improvement by avoiding using
`lowercase` as much as possible in the parsing. In order to achieve the
fastest performace we only call `lowercase` when we have not cached the
provided case-sensitive version of the string.

```julia
julia> @benchmark DateTime("Sat, 12 Nov 2016 07:45:36", Base.Dates.RFC1123Format)  # cached
BenchmarkTools.Trial:
  memory estimate:  192.00 bytes
  allocs estimate:  6
  --------------
  minimum time:     461.863 ns (0.00% GC)
  median time:      468.622 ns (0.00% GC)
  mean time:        539.984 ns (1.79% GC)
  maximum time:     8.893 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     197
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark DateTime("sat, 12 Nov 2016 07:45:36", Base.Dates.RFC1123Format)  # cached
BenchmarkTools.Trial:
  memory estimate:  192.00 bytes
  allocs estimate:  6
  --------------
  minimum time:     460.939 ns (0.00% GC)
  median time:      478.338 ns (0.00% GC)
  mean time:        553.019 ns (1.83% GC)
  maximum time:     8.452 μs (91.02% GC)
  --------------
  samples:          10000
  evals/sample:     197
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark DateTime("sAt, 12 Nov 2016 07:45:36", Base.Dates.RFC1123Format)  # uncached
BenchmarkTools.Trial:
  memory estimate:  464.00 bytes
  allocs estimate:  10
  --------------
  minimum time:     698.469 ns (0.00% GC)
  median time:      733.701 ns (0.00% GC)
  mean time:        897.745 ns (6.67% GC)
  maximum time:     32.580 μs (96.06% GC)
  --------------
  samples:          10000
  evals/sample:     147
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

omus commented:
@shashi using non-numeric slots `[uUeE]` slow down the performance quite a bit. Your current implementation skips any checks on the day of week which means `"Foo, 12 Nov 2016 07:45:36"` will parse correctly with the date format `DateFormat("e, dd u yyyy HH:MM:SS")`. I feel we should verify that the day of week is valid which will come at the cost of performance (seems to add ~250ns). What do others think?

ViralBShah commented:
Maybe check it by default, but have a way to skip it with an optional argument?

ViralBShah commented:
One thing that I heard in my recent trip meeting lots of Julia users is that people *really* want the fastest date time parsing possible, and having a way to disable certain checks may be a reasonable thing to do, while having a strict default.

Out of curiosity, how much of the total parsing time is the 250ns check, in terms of percentage?

omus commented:
@ViralBShah the validation ends up being ~35% of the total parsing time. I'll see if I can get the time down yet.

omus commented:
I'll note that I did a by custom code version of ISO8601 and was getting even better performance. For users who have a non-standard format and want the fastest speed possible that seems to be the way to go.

shashi commented:
Yes adding an "ignore" specifier is easy. I don't know which character to use for it though. `?` maybe makes sense? I didn't go ahead and do it because a few other questions came to mind, like can we allow regular expressions to be spliced into date time format? The Delim type allows this, I'm not sure actual date fields (DatePart) need this. A possible syntax is `datetime"y$(r"myregex")mmdd"`.

@omus does the 250ns also include checking to see if the parsed date actually is that day? If not, in theory a compiled regular expression check should be faster than a dict lookup.

shashi commented:
the performance slowdown with `uUeE` is expected. Will be nice to minimize them though. Note that the previous behavior was also to ignore day of week strings.

omus commented:
Currently the code is just verifying that the day of week is a day of week and doesn't verify that the day of week corresponds to the rest of the datetime.

omus commented:
I tracked down the slow down with `uUeE` to be from the use of `lowercase`. I've managed to eliminate this call in most scenarios which has produced a noticeable speed improvement for parsing of month/day-of-week words.

omus committed:
Support returning abstract types from tryparse_internal

omus committed:
Fix overflow issues on 32-bit Julia

omus committed:
Refactor some tests

Update testcases

Avoid passing test silently if no exception thrown

omus committed:
Dispatch on custom parser for improved performance

```julia
julia> s = "2010-10-11T12:13:14.15"
"2010-10-11T12:13:14.15"

julia> df_default = Base.Dates.DateFormat("yyyy-mm-ddTHH:MM:SS.s")
dateformat"yyyy-mm-ddTHH:MM:SS.s"

julia> df_custom = Base.Dates.DateFormat("yyyy-mm-dd\\THH:MM:SS.s")
dateformat"yyyy-mm-ddTHH:MM:SS.s"

julia> @benchmark parse(DateTime,$s,$df_default)
BenchmarkTools.Trial:
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     127.267 ns (0.00% GC)
  median time:      135.160 ns (0.00% GC)
  mean time:        147.785 ns (0.00% GC)
  maximum time:     2.450 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     886
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark parse(DateTime,$s,$df_custom)
BenchmarkTools.Trial:
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     71.947 ns (0.00% GC)
  median time:      74.662 ns (0.00% GC)
  mean time:        82.548 ns (0.00% GC)
  maximum time:     2.529 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     976
  time tolerance:   5.00%
  memory tolerance: 1.00%

```

kshyatt commented:
@nalimilan and @quinnj would one of you mind reviewing this?

tkelman commented:
according to https://github.com/JuliaLang/julia/pull/19519#issuecomment-267779592 @omus has more improvements pending, due monday?

shashi commented:
Merged @omus's #19519 into this. This is ready for further review.

shashi commented on base/dates/parse.jl:
```diff
@@ -0,0 +1,184 @@
+
+    field_defaults = slot_defaults(T)
+    field_order = tuple(order...)
+    tuple_type = slot_types(T)
```
@vtjnash Are these changes to the generated function fine by you?

vtjnash replied:
It's not very flexible (you won't be able to provide new definitions for these functions, such as `slot_defaults `), but I think it's valid.

However, I don't think `NTuple{N}` is valid (here and elsewhere), since with https://github.com/JuliaLang/julia/pull/18457 that will mean "a tuple with N elements of the *same* type"

omus replied:
I assume that will also be an issue for `@ntuple`?

tkelman replied:
this is still a major issue, isn't it?

shashi replied:
@omus why would it be an issue for `@ntuple`? It's a macro which writes N expressions in a tuple literal afaict...

omus replied:
@shashi it would only be an issue if `@ntuple` was forced to return an `NTuple`. Since it returns a `Tuple` we should be fine.

vtjnash replied:
looking through Jeff's PR, you just need to replace `NTuple{N}` with `NTuple{N, Any}` to keep it working

shashi commented on base/dates/parse.jl:
```diff
@@ -115,3 +115,70 @@ function rfixwidth(num, n)
     length(s) < n ? lpad(s, n, 0) : s[end - n + 1:end]
 end
 
+function Base.parse(::Type{DateTime}, s::AbstractString, df::DateFormat{Symbol("yyyy-mm-dd\\THH:MM:SS.s")})
```
This is what I was trying to avoid by having the number of times a token is repeated and fixwidth in the type parameters - you could just dispatch on the format object's `typeof`. This is fine too, and probably better.

ViralBShah commented:
We probably need squashing of commits (with somewhere between 1 and 53 commits), preserving authorship and meaning.

omus commented:
Feel free to squash most of my commits together however you like. I would recommend squashing 6c000c55b297b69eabbead4bb5e055db8776e12b and  08182294334b1bf98ad259f0329afe978527f66d together (custom parser implementation) and keeping the result as a separate commit. I have a feeling this functionality could be controversial.

shashi committed:
Speed up date time parsing using a generated function.

Introduces a new DateFormat type and tryparsenext generic function. An efficient parser is now generated for every format string.

shashi committed:
Updates to date locale interface

- remove symbol param from DateLocale type
- DateLocale now contains all required info about a locale
- Introduce Dates.LOCALES Dict{String, DateLocale} to tack
  on new locales by name.

Minor corrections to locale changes (by Curtis Vogt, squashed)

Default locale now uses DateLocale object (by Curtis Vogt, squashed)

omus commented:
Can/should this be added to the 0.6.0 milestone?

yuyichao commented:
Doesn't seem blocking at all so no. (It can certainly be merged before 0.6 feature freeze or even after if it's just a performance improvement)

omus commented:
Could someone run the nanosoldier benchmarks? Thanks

andreasnoack commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

JeffBezanson commented:
The priority of this is much higher than it might seem at first. The current date parser is so slow that it's a bug. It entirely prevents using julia to handle CSV files with dates (which is lots of them).

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19545#issuecomment-269541996) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/af09213adf9c68ccb2d8a36a28a27c5df29fdb36/d144f67_vs_309657f/report.md). cc @jrevels

andreasnoack commented:
The speedups for parsing look quite remarkable but it also looks like there are genuine problems in some `Dates` related benchmarks. They allocate a lot more and are much slower.

quinnj commented:
Why would this affect other Dates benchmarks? It seems like that may be some kind of noise or mal-configuration.

Sorry to be so silent on this; I've been travelling for a week or so and probably won't have time to really review for another couple days or so; if people feel like it's generally good to go, feel free to merge; otherwise, I'll try and do a real dig-in to the code here and do a proper review.

andreasnoack commented:
In this PR, conversion of a `Date` to a `String` is [this method](https://github.com/JuliaLang/julia/blob/sh/fast_date_parsing/base/dates/io.jl#L415-L417) but used to be [this method](https://github.com/JuliaLang/julia/blob/master/base/dates/io.jl#L17-L23). The allocation is 3x up in one of the benchmarks so it is not noise.

quinnj commented:
ok, but still IO related, which is what this PR is all about. I thought there was some non-IO benchmarks showing negative results.

ViralBShah commented:
I believe we should revisit performance after the strings PR is merged as well. For now, it would be good to get this merged. And @quinnj should do a proper review when he can, even if after the merge.

shashi commented:
I'm trying to improve the printing speeds, I might have reclaimed the speed but there should be room for improvement.

shashi committed:
make DateTime printing faster

shashi commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19545#issuecomment-269614287) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2bf3cb1f1542a32cb4946bc56a43e883d2fcca68/e6ffc04_vs_b561cfb/report.md). cc @jrevels

tkelman commented on base/dates/io.jl:
```diff
@@ -1,244 +1,219 @@
 
 """
-    DateTime(dt::AbstractString, format::AbstractString; locale="english") -> DateTime
+    DateFormat(format::AbstractString, locale="english", default_fields=(1,1,1,0,0,0,0)) -> DateFormat
```
what is `default_fields` doing?

omus replied:
Should be removed from the docstring. Was needed in an earlier revision of this PR.

tkelman commented on base/dates/parse.jl:
```diff
@@ -0,0 +1,172 @@
+    end
+end
+
+function reorder_args{Nv, Ni}(val::NTuple{Nv}, idx::NTuple{Ni}, default::NTuple{Ni}, valid_till)
```
could also use some comments on what this reorder_args business is doing

tkelman commented on base/dates/parse.jl:
```diff
@@ -0,0 +1,172 @@
+
+function Base.tryparse{T<:TimeType}(::Type{T}, str::AbstractString, df::DateFormat)
+    nt = tryparse_internal(T, str, df, false)
+    Nullable{T}(isnull(nt) ? nothing : T(nt.value...))
```
same comment as above about `Nullable{T}(nothing)`

tkelman commented on base/dates/parse.jl:
```diff
@@ -0,0 +1,172 @@
+    th = tm = ts = tms = Int64(0)
+
+    nv, i = tryparsenext_base10(s,i,end_pos,1)
+    dy = isnull(nv) ? (@goto error) : get(nv)
```
does this duplicate the null check?

tkelman commented on base/dates/parse.jl:
```diff
@@ -0,0 +1,172 @@
+    tms *= 10 ^ (3 - (j - i))
+    i = j
+
+    i > end_pos || @goto error
```
not much need to have this use `i` instead of `j` is there? given the rest of the lines that look like this are `@goto done`, this one is different with `@goto error` so may as well also be different with `j > end_pos`, then don't need the `i = j`

omus replied:
Very true. I originally did this so I could remove the millisecond chunk without having to modify that check. I'm fine with your suggestion.

tkelman commented on base/dates/query.jl:
```diff
@@ -29,31 +100,31 @@ dayofweek(dt::TimeType) = dayofweek(days(dt))
 """
-dayname(dt::TimeType;locale::AbstractString="english") = VALUETODAYOFWEEK[locale][dayofweek(dt)]
+function dayname(dt::TimeType;locale::AbstractString="english")
+   dayname(dayofweek(dt); locale=locale)
```
under-indented (same below with `dayabbr`)

tkelman commented on test/dates/io.jl:
```diff
@@ -267,8 +291,7 @@ f = "y m d"
 @test_throws ArgumentError Dates.Date("# 1 1 32",f)
-# can't find 1st space delimiter,s o fails
-@test_throws ArgumentError Dates.Date("1",f)
+@test Dates.Date("1",f) == Dates.Date(1)
```
so the new parsing is more permissive of partial matches? should that be mentioned in news or somewhere else?

omus replied:
There was a special case in the original code that said: if the first slot is succeeded by a delimiter then the first slot and the delimiter is required.

Before PR
```julia
julia> Dates.Date("1", "y-m-d")
ERROR: ArgumentError: Delimiter mismatch. Couldn't find first delimiter, "-", in date string
 in parse(::String, ::Base.Dates.DateFormat) at ./dates/io.jl:169
 in Date(::String, ::String) at ./dates/io.jl:277

julia> Dates.Date("1-", "y-m-d")
0001-01-01

julia>  Dates.Date("1", "ymd")
0001-01-01
```

After PR:
```julia
julia> Dates.Date("1", "y-m-d")
0001-01-01

julia> Dates.Date("1-", "y-m-d")
0001-01-01

julia> Dates.Date("1", "ymd")  # Note: currently there is a bug with this.
0001-01-01
```

omus replied:
Also, the original code was inconsistent with this rule as it trimmed whitespace automatically:

Before PR:
```
julia> Dates.Date("1 ", "y m d")  # Delimiter removed when removing whitespace
ERROR: ArgumentError: Delimiter mismatch. Couldn't find first delimiter, " ", in date string
 in parse(::String, ::Base.Dates.DateFormat) at ./dates/io.jl:169
 in Date(::String, ::String) at ./dates/io.jl:277
```

tkelman replied:
Thanks for the answers here - sounds reasonable. The behavior changes make this not a backport candidate though.

omus replied:
We have some options if people want this functionality in 0.5. The simplest would be to backport the dispatch for custom parsers which would allow us to get the performance improvement for ISO 8601 formatted strings.

tkelman commented on test/dates/io.jl:
```diff
@@ -77,7 +94,7 @@ c3 = "96:1:01"
 @test Dates.format(dt-Dates.Day(14),"yy:m:dd") == c3
 c4 = "1996:15:01 # random comment"
-@test Dates.DateTime(c4,f) == dt
+@test_throws ArgumentError Dates.DateTime(c4,f)
```
is this breaking then, or was allowing comments before unintentional? can that capability be recreated now with a fancier date format if someone needs it?

omus replied:
Possibly unintentional. I believe the original implementation just ignored the remainder of a string when had collected all the slots indicated by the format.

omus replied:
Personally, I think removal of comments should probably happen as a pre-processing step if it is needed.

shashi replied:
I agree that it should be a pre-processing step.

I think we should merge this now and then open a PR against 0.5 branch with changes that make it possible to backport this.

tkelman commented on test/dates/io.jl:
```diff
@@ -198,7 +215,7 @@ s = "/1996/1/15"
 @test Dates.DateTime(s,f) == dt
 @test Dates.format(dt,f) == s
-@test Dates.DateTime("1996/1/15",f) == dt
+@test_throws ArgumentError Dates.DateTime("1996/1/15",f)
```
was not being picky about the leading slash a bug or intentional before?

shashi replied:
My guess is it was an artefact of how it was implemented. @quinnj

omus committed:
Address review comments (#19764)

Fix under-indented function bodies

Fix bug where default wasn't used

Avoid duplicate isnull checks

Remove unnecessary assignment

Add a test for whitespace delimiter

Test uncovered an issue that existed prior to #19545

Expand inline Nullable construction

Corrections to documentation

Switch to using unsafe_get

omus committed:
Fix bad usage of NTuple

More NTuple fixes (amended it to use NTuple{N} (known to be of Ints) in
reorder_args )

shashi commented on base/dates/io.jl:
```diff
@@ -356,7 +356,8 @@ Parse a date from a date string `dt` using a `DateFormat` object `df`.
 
-@generated function format{S,N}(io::IO, dt::TimeType, fmt::DateFormat{S,NTuple{N}})
+@generated function format{S,T}(io::IO, dt::TimeType, fmt::DateFormat{S,T})
+    N = length(T.parameters)
```
Why is this bad usage? Also, this could be `nfields`

shashi replied:
Ah I see now, I just saw Jameson's comment above.

shashi commented on base/dates/parse.jl:
```diff
@@ -50,8 +54,9 @@ function gen_exception(tokens, err_idx, pos)
-    ntuple(Val{Ni}) do i
+function reorder_args(val::Tuple, idx::Tuple, default::Tuple, valid_till::Integer)
+    # Note: idx and default have the same length
+    ntuple(length(idx)) do i
```
This allocates and is slower...?

shashi replied:
@vtjnash any other way to facilitate the compiler figuring out the length of the tuple? I presume that uncertainty is why it needs to box this tuple?

shashi replied:
Oh never mind, I know that idx is always `NTuple{N,Int}`

omus replied:
We should be able to switch `idx` back to an `NTuple` to avoid the length call.

shashi commented on base/dates/parse.jl:
```diff
@@ -1,7 +1,8 @@
-    token_types = Type[dp <: DatePart ? SLOT_RULE[first(dp.parameters)] : Void for dp in df.parameters[2].parameters]
+@generated function tryparse_internal{T<:TimeType,S,F}(::Type{T}, str::AbstractString, df::DateFormat{S,F}, raise::Bool=false)
+    token_types = Type[dp <: DatePart ? SLOT_RULE[first(dp.parameters)] : Void for dp in F.parameters]
+    N = length(F.parameters)
```
I preferred to keep minimal logic in the `@generated` function as it makes it easier to understand...

tkelman commented on base/dates/io.jl:
```diff
@@ -1,16 +1,45 @@
+
 """
-    AbstractDateToken
+     AbstractDateToken
```
this was right before

tkelman commented on base/dates/io.jl:
```diff
@@ -1,16 +1,45 @@
-    tryparsenext(t::DatePart, str, i, len, [locale])
-    format(io, t::AbstractDateToken, dt, [locale])
 """
+
```
shouldn't have excess lines between a docstring and what it's attached to

tkelman commented on base/dates/io.jl:
```diff
@@ -99,13 +128,14 @@ for (tok, fn) in zip("eE", [dayabbr, dayname])
+
+    # the last `n` digits of `y`
+    # will be 0 padded if `num` has less than `n` digits
+    str = dec(num,n)[end-(n-1):end]
```
y, not num? latter isn't defined here (does this cause a test failure? if not, needs coverage)

shashi committed:
add comments as requested in review

remove fixwidth function which was used only once. Inline the code now.

shashi commented:
Tony's comments have been addressed (Github isn't letting me reply to them in-line). I've rebased the branch down to 21 commits. Regressions reported by Nanosoldier are unrelated.

omus commented:
I've found one issue with `tryparse_internal` having a type instability when using custom `slot_types`. It only matters for those extending the code (like me) but doesn't impact `Date` or `DateTime` parsing performance. Should be able to take care of it today.

omus commented:
The type instability I pointed out is proving harder to remove without impacting performance. I can address this issue as a fix later if I find a solution. I'm good with proceeding as is.

omus commented on base/dates/query.jl:
```diff
@@ -72,8 +143,10 @@ For the day of week of `dt`, returns which number it is in `dt`'s month. So if t
-    d < 15 ? 2 : d < 22 ? 3 : d < 29 ? 4 : 5)
+function dayofweekofmonth(dt::TimeType)
+    d = day(dt)
+    return d < 8 ? 1 :d < 15 ? 2 : d < 22 ? 3 : d < 29 ? 4 : 5
```
@shashi I forgot to address this comment. Can you take care of it?

shashi replied:
Done.

tkelman commented:
Some of the early intermediate commits fail tests - specifically 1050c7ad18c1a9ca84e29696e26322e4f7f2ffc1, should that be squashed together with the following 685bb9876e51058cd0d668d13717e27e7ddefb3d to avoid tripping up future bisects?

tkelman commented:
eb94a59955423be3a1f205f0d680a3a85f08a2f7 also fails to build

shashi committed:
Refactor

Organizing methods

Refactor rfixwidth to be consistent with minwidth

Minor refactor to DateFormat constructor

shashi commented:
Rebased to have no intermediate failing commits. Waiting for CI.

quinnj reviewed (changes requested):
Overall looks great! Definitely some great improvements here. I added a few comments here and there, but most are more cosmetic. Overall, it's encouraging how little the actual tests had to change. Great work @omus and @shashi!

shashi commented on base/dates/parse.jl:
```diff
@@ -0,0 +1,197 @@
+    T(unsafe_get(nt)...)
+end
+
+@inline function tryparsenext_base10(str::AbstractString, i::Int, len::Int, min_width::Int=1, max_width::Int=0)
```
Yeah, I plan to include that in a later PR. Will remove this then.

shashi commented on base/dates/io.jl:
```diff
@@ -1,244 +1,248 @@
+end
+
+function format(io, d::DatePart{'s'}, dt)
+    write(io, string(millisecond(dt)/1000)[3:end])
```
Would creating a substring avoid a copy?

quinnj replied:
I think that would work; in Feather, I resorted to always writing a `view` of the underlying bytes, but a Substring should work too.

quinnj commented on base/dates/parse.jl:
```diff
@@ -0,0 +1,197 @@
+    end
+end
+
+function Base.parse(::Type{DateTime}, s::AbstractString, df::DateFormat{Symbol("yyyy-mm-dd\\THH:MM:SS.s")})
```
Yeah, you're right.

shashi commented on base/dates/io.jl:
```diff
@@ -261,8 +265,99 @@ Characters not listed above are normally treated as delimiters between date and
+# Standard formats
+const ISODateTimeFormat = DateFormat("yyyy-mm-dd\\THH:MM:SS.s")
+const ISODateFormat = DateFormat("yyyy-mm-dd")
+const RFC1123Format = DateFormat("e, dd u yyyy HH:MM:SS")
```
Currently we don't have parsing for time zones in Base do we?

omus replied:
Time zone parsing is in https://github.com/JuliaTime/TimeZones.jl

shashi replied:
So it works by looking at `Delim{'Z'}`? When someone is not using TimeZones, Base parses till before Z and everything keeps working? Kinda feels like a hack, but if it works... Feel free to push this change if you see it makes sense.

shashi committed:
Address review comments

use SubString instead of allocating
default parse method for Date and DateTime
smaller case french months and days of week
doc fixes

shashi commented:
`@nanosoldier runbenchmarks("dates", vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19545#issuecomment-272617225) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/90127f7a2f72480d1afe2fa8d25891d8461eab78/869e3d6_vs_8afb74d/report.md). cc @jrevels

shashi committed:
Merge remote-tracking branch 'origin/master' into sh/fast_date_parsing

shashi commented:
Just realized the slowdowns are coming because `jb/fasterstring` was merged before the benchmark ran! :) Using SubString does not seem to make much of difference to the timing here.

Let's try this again: @nanosoldier `runbenchmarks("dates", vs = ":master")`

tkelman commented:
That shouldn't matter, nanosoldier should be running on the merge commit

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19545#issuecomment-273657816) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/b3480d9243c0a951cf4146686f2b265de10ef3da/c64d561_vs_0dfdc2c/report.md). cc @jrevels

tkelman commented:
The regressions are real, but we should get this in before feature freeze. @omus @shashi have you looked into fixing them?

shashi commented:
I tried reverting the change. On my machine there's barely any change. I don't know how https://github.com/JuliaLang/julia/pull/19545#issuecomment-269637545 gave a green tick for these benchmarks.

tkelman commented:
I'm not sure, but it's possible this might be hitting problems related to #20025 ?

tkelman commented:
That good set of results was just before we merged the llvm 3.9 upgrade, so newer llvm changes the performance characteristics of the patches here w.r.t. those benchmarks.

StefanKarpinski commented:
Wait, so what's the status here? The string change is so kickass that it makes these speed ups irrelevant?

shashi commented:
@StefanKarpinski No, there's a regression in converting DateTime to string which I'm trying to track down. Parsing is still much much faster with this.

shashi commented on base/dates/io.jl:
```diff
@@ -324,22 +433,41 @@ generate the string "1996-01-15T00:00:00" you could use `format`: "yyyy-mm-ddTHH
+        dateformat"YYYY-mm-dd\THH:MM:SS"
+    else
+        dateformat"YYYY-mm-dd\THH:MM:SS.s"
+    end
```
@omus @quinnj is it okay if we always use `"YYYY-mm-dd\THH:MM:SS.s"` here? This type instability adds to the printing slow down. Users wanting to ignore milliseconds can specify `dateformat"YYYY-mm-dd\THH:MM:SS"`

quinnj replied:
Fine by me.

shashi replied:
Oh, wait. this is not really a problem...

omus replied:
Glad it's not a problem. If we were to change this I think I would prefer `"YYYY-mm-dd\THH:MM:SS.sss"`

shashi committed:
[ci skip] further speed up date time printing

shashi commented:
@nanosoldier `runbenchmarks("dates", vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19545#issuecomment-274498070) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/5b0f2f56345a90a3697a4f413fbda39aa14a39d3/7784d3c_vs_5398799/report.md). cc @jrevels

quinnj commented:
`DateTime` printing looks good! Do we just need to apply the same changes to `Date`?

omus commented on base/dates/io.jl:
```diff
@@ -1,244 +1,263 @@
+    else
+        l = endof(str)
+        write(io, SubString(str,l-(n-1),l))
     end
```
Should probably save `endof(str)` before the if.

tkelman commented on base/dates/io.jl:
```diff
@@ -324,22 +450,48 @@ generate the string "1996-01-15T00:00:00" you could use `format`: "yyyy-mm-ddTHH
+    if millisecond(dt) == 0
+        format(dt, dateformat"YYYY-mm-dd\THH:MM:SS", 24)
+    else
+        format(dt, dateformat"YYYY-mm-dd\THH:MM:SS.s", 26)
```
are milliseconds in a DateTime only ever a single digit?

shashi replied:
No, `s` stands for any number of digits e.g. `s` gives `.5` or `.05` or `.005` while `.sss` would give `.500`, `.050`.

tkelman replied:
so is the 26 a minimum?

shashi replied:
No it's usually 23 chars, I added 3 more bytes to keep this going fast even when the year is `-10000` ;) I'm not sure if this really matters, doesn't julia allocate the buffer backing an array in multiples of 16 bytes or something?

tkelman replied:
Not sure. As long as this is just a rough-guess buffer preallocation and things work fine if the actual output is longer or shorter then I'll just go with whatever the benchmarks say.

Commit:
Merge remote-tracking branch 'origin/master' into sh/fast_date_parsing_merge

shashi commented:
Date formatting (to arbitrary formats) is 5x faster than before. So in a bit of a cop out, I changed `string(::Date)` to be the same as before. I guess the speed up mainly comes from not allocating an IOBuffer object.

@nanosoldier `runbenchmarks("dates", vs = ":master")`

tkelman commented on base/dates/io.jl:
```diff
@@ -339,22 +469,54 @@ generate the string "1996-01-15T00:00:00" you could use `format`: "yyyy-mm-ddTHH
+    end
+end
+
+function Base.string(dt::Date)
```
do you want to delete the now-duplicate one at line 46 then?

shashi replied:
Oh, yes, I should overlooked that when resolving merge conflicts.

tkelman replied:
there's still a duplicate method here

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19545#issuecomment-274721383) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/8ac31a083e02803985556c667d35985d0edbac2a/c8d3875_vs_06fa32c/report.md). cc @jrevels

tkelman commented on base/dates/io.jl:
```diff
@@ -1,30 +1,47 @@
+format the `tok` token from `dt` and write it to `io`. The formatting can
+be based on `locale`.
+
+all subtypes of `AbstractDateToken` must define this method in order
```
First word in a sentence should be capitalized
